captum/optim/_core/optimization.py

```diff
@@ -54,20 +51,22 @@ def __init__(
                         consumed by the model.
             transform (nn.Module, optional):  A module that transforms or preprocesses
                         the input before being passed to the model.
-            target_modules (iterable of nn.Module):  A list of targets, objectives that
-                        are used to compute the loss function.
             loss_function (callable): The loss function to minimize during optimization
                         optimization.
             lr (float): The learning rate to use with the Adam optimizer.
         """
         self.model = model
-        self.hooks = ModuleOutputsHook(target_modules)
+        # Grab targets from loss_function
+        if hasattr(loss_function.target, "__iter__"):
+            self.hooks = ModuleOutputsHook(loss_function.target)
+        else:
+            self.hooks = ModuleOutputsHook([loss_function.target])
         self.input_param = input_param or NaturalImage((224, 224))
+        self.input_param.to(next(self.model.parameters()).device)
         self.transform = transform or torch.nn.Sequential(
             RandomScale(scale=(1, 0.975, 1.025, 0.95, 1.05)), RandomSpatialJitter(16)
         )
         self.loss_function = loss_function

```

Comment:
If the code fails or any reason after creating the hooks, we end up with the same issue that the `optimize` class function had. If you add a line like raise error to the `init` function, then you can reliably get the error. I think that users may be faced with the issue while doing normal things as I've encountered it while testing transforms and other stuff.

We resolved the issue in `optimize` by using this:

```
         finally:
            self.cleanup()
```

The init function appears to be the only other place that this issue can occur.

Reply:
So actually it looks like the same error will happen if there are any errors that occur between `InputOptimization` and running the `optimize` function.
```
    obj = opt.InputOptimization(model, loss_fn, image, transforms)
    #raise Error
    history = obj.optimize(opt.optimization.n_steps(128, True))
```

Then without forcing an error:

```
/content/captum/captum/optim/_core/loss.py in __call__(self, targets_to_values)
    240     def __call__(self, targets_to_values: ModuleOutputMapping) -> torch.Tensor:
    241         activations = targets_to_values[self.target]
--> 242         assert activations is not None
    243         assert self.channel_index < activations.shape[1]
    244         assert len(activations.shape) == 4  # assume NCHW

AssertionError: 
```

I'm actually not sure if we can fix this issue?

Reply:
We have to leave solving this issue for a future PR